File: leetcode/930.binary-subarrays-with-sum.cpp
/*
 * @lc app=leetcode id=930 lang=cpp
 *
 * [930] Binary Subarrays With Sum
 */

// @lc code=start
class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        return count(nums, goal) - count(nums, goal-1);        
    }
    int count(vector<int>& nums, int goal) {
        int left = 0;
        int array_count=0;
        int sum=0;

        for (int right = 0; right<nums.size(); right++) {
            sum+=nums[right];
           
            while (left<=right && sum>goal) {
                sum-=nums[left];
                left++;
            }
            array_count += (right - left + 1);
        }

      
        return array_count;
    }
};
// @lc code=end

File: leetcode/930.binary-subarrays-with-sum.py
#
# @lc app=leetcode id=930 lang=python3
#
# [930] Binary Subarrays With Sum
#

# @lc code=start
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        def count_array(nums: List[int], goal: int):
            left=0
            count=0
            sum=0
        
            for right, num in enumerate(nums):
                sum+=num
                while (left<=right and sum>goal ):
                    sum-=nums[left]
                    left+=1
                
                count+= (right - left + 1)
            
            return count
    
        return count_array(nums, goal) - count_array(nums, goal-1)
    
    
    
        
        
        
# @lc code=end

